workingCopyService: take options alongside each argument

isidorn opened the pull request:
Refs: https://github.com/microsoft/vscode/issues/111867

This PR is work in progress.
I just made sure tests pass and vscode can start nicely.

Feedback on naming is welcome.
I left 2 todo@isidor in code where I would appreciate advice

I am pretty happy with how it turnet out, only not sure about the whole `SourceTargetPairWithOptions`.

bpasero commented on src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts:
```diff
@@ -31,6 +31,13 @@ export interface SourceTargetPair {
 }
 
+export interface SourceTargetPairWithOptions extends SourceTargetPair {
+	options?: {
```
This could get some JDoc comment what the purpose and impact of these options is.

bpasero commented on src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts:
```diff
@@ -57,15 +64,34 @@ export interface IWorkingCopyFileOperationParticipant {
+	options?: {
+		useTrash?: boolean;
+		recursive?: boolean;
+		undoRedoGroupId?: number;
```
I wonder if there is a way to extract a generic interface for `undoRedoGroupId` and `isUndoing` and then use it for all operations and the `SourceTargetPairWithOptions`?

bpasero commented on src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts:
```diff
@@ -284,36 +310,38 @@ export class WorkingCopyFileService extends Disposable implements IWorkingCopyFi
 		// validate move/copy operation before starting
-		for (const { source, target } of files) {
-			const validateMoveOrCopy = await (move ? this.fileService.canMove(source, target, overwrite) : this.fileService.canCopy(source, target, overwrite));
+		for (const operation of operations) {
```
Could benefit from some destructuring when you loop over operations? Try with `for (const { file: { source, target }, options } of operations) {` and you get every property you need in one go.

bpasero commented on src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts:
```diff
@@ -284,36 +310,38 @@ export class WorkingCopyFileService extends Disposable implements IWorkingCopyFi
 		// file operation participant
-		await this.runFileOperationParticipants(files, move ? FileOperation.MOVE : FileOperation.COPY, options?.undoRedoGroupId, options?.isUndoing, token);
+		const files = operations.map(o => o.file);
+		// TODO@isidor should the options be passed here
```
Any reason why not?

bpasero reviewed (changes requested):
Left some feedback inline. For consistency I would also add similar operation objects for the `create` and `createFolder` methods 👍

jrieken commented on src/vs/workbench/api/browser/mainThreadFileSystemEventService.ts:
```diff
@@ -74,8 +74,8 @@ export class MainThreadFileSystemEventService {
-			async participate(files: SourceTargetPair[], operation: FileOperation, undoRedoGroupId: number | undefined, isUndoing: boolean | undefined, timeout: number, token: CancellationToken) {
-				if (isUndoing) {
+			async participate(files: SourceTargetPairWithOptions[], operation: FileOperation, timeout: number, token: CancellationToken) {
+				if (files.some(f => f.options?.isUndoing)) {
```
My feeling is that some options, like `isUndoing`, aren't part of the `SourceTargetPairWithOptions`. Like the operation is undoing 3 creations etc, so isUnding and the undoRepo groups belong is a single options bag if you ask me.

isidorn replied:
This suggestoin makes sense to me. I will go down this path.
@bpasero fyi

isidorn commented:
I have addressed your comments. I am now reusing the `IFileOperationUndoRedoInfo` between `WorkingCopyFileServie` and `BulkEdits`. I like how it looks now, let me know what you think

Still two more things to do
- [ ] add jsdoc
- [ ] update `create` and `createFolder` methods

isidorn commented:
I have also tackled the `create` and `createFolder`. Note that I am calling all the creation in parallel for now in these two methods.
Feedback on jsdoc is appreciated since I am not actually sure how `isUndoing` is being used in the layers below.
I ran tests and tried various explorer operations and they seem ok.

Let me know what you think, thanks!

isidorn commented on src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts:
```diff
@@ -344,19 +377,19 @@ export class WorkingCopyFileService extends Disposable implements IWorkingCopyFi
-		for (const resource of resources) {
-			const validateDelete = await this.fileService.canDelete(resource, options);
+		for (const operation of operations) {
+			const validateDelete = await this.fileService.canDelete(operation.resource, { recursive: operation.recursive, useTrash: operation.useTrash });
```
Instead of creating a new options bag I could just pass the `operation` object
@bpasero let me know what you prefer.
Samo on line `409`

bpasero replied:
I think it is fine to create a new option bag, the way it is now 👍 . I often prefer to explicitly call out how a method is called to make it clear when reading the code.

bpasero reviewed (changes requested):
Overall looks already pretty good to me, just one feedback I have: you now allow to create multiple files or folders in the working copy file service. I am not sure that is needed, so maybe we go back to just single operation there (keep the `ICreateFileOperation`, and `ICreateOperation` you added, those make good sense).

The reason I feel we don't need it is: we want to support it, we also should make sure that the related participants and events for create get all the files at once, similar to how we do it in move and copy. Currently it seems you still iterate over each file and create it, but the purpose of allowing multiple operations at once was specifically to only send 1 event per array of files. 

Since we don't really have a usecase for that currently (to my knowledge), I suggest to go back how it was.

isidorn commented:
@bpasero thanks a lot for feedback. I also think it makes more sense to not allow multi operation passing, but I misunderstood you in our private chat and thus I added it..
Once I tackle that I plan to merge this in.

Since you are in next week please keep an eye open if I broke something with this :) But I am fairly confident.

isidorn commented:
@bpasero Ok you just said in standup you will be out
So just @jrieken keep this PR in mind if some file features get broken from source.
